File: Sources/Services/StashAPI.swift
import Foundation

enum StashAPIError: LocalizedError {
    case badKey
    case http(Int)

    var errorDescription: String? {
        switch self {
        case .badKey:
            return "That API key was rejected. Check it in the Windows app under Settings › Cloud Sync › API Keys."
        case .http(let code):
            return "The stash service returned an error (HTTP \(code))."
        }
    }
}

/// Thin client for the My Crafty Stash cloud API (Azure Static Web Apps).
/// Auth is the same mcs_ API key the Windows desktop app uses, sent as an
/// X-Api-Key header.
struct StashAPI {
    static let defaultBaseURL = URL(string: "https://wonderful-meadow-0a1a40110.7.azurestaticapps.net")!

    var baseURL: URL = StashAPI.defaultBaseURL
    var apiKey: String

    func whoami() async throws -> Whoami {
        try await get("/api/whoami")
    }

    func items(page: Int = 1, perPage: Int = 48, search: String? = nil, type: String? = nil) async throws -> ItemsPage {
        var query = [
            URLQueryItem(name: "page", value: String(page)),
            URLQueryItem(name: "perPage", value: String(perPage)),
        ]
        if let search, !search.isEmpty { query.append(URLQueryItem(name: "search", value: search)) }
        if let type, !type.isEmpty { query.append(URLQueryItem(name: "type", value: type)) }
        return try await get("/api/items", query: query)
    }

    private func get<T: Decodable>(_ path: String, query: [URLQueryItem] = []) async throws -> T {
        var comps = URLComponents(url: baseURL.appendingPathComponent(path), resolvingAgainstBaseURL: false)!
        if !query.isEmpty { comps.queryItems = query }

        var req = URLRequest(url: comps.url!)
        req.setValue(apiKey, forHTTPHeaderField: "X-Api-Key")

        let (data, response) = try await URLSession.shared.data(for: req)
        let status = (response as? HTTPURLResponse)?.statusCode ?? 0
        switch status {
        case 200:
            return try JSONDecoder().decode(T.self, from: data)
        case 401:
            throw StashAPIError.badKey
        default:
            throw StashAPIError.http(status)
        }
    }
}

File: MyCraftyStash/Services/StashApi.cs
using System.Net;
using System.Net.Http.Json;
using System.Text.Json;
using MyCraftyStash.Models;

namespace MyCraftyStash.Services;

public class StashApiException : Exception
{
    public StashApiException(string message) : base(message) { }
}

/// <summary>
/// Thin client for the My Crafty Stash cloud API (Azure Static Web Apps).
/// Auth is the same mcs_ API key the Windows desktop app uses, sent as an
/// X-Api-Key header. The key is read from <see cref="StashSession"/> per
/// request so sign-in/out never needs a new client.
/// </summary>
public class StashApi
{
    public const string DefaultBaseUrl = "https://wonderful-meadow-0a1a40110.7.azurestaticapps.net";

    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNameCaseInsensitive = true,
    };

    private readonly HttpClient _http = new() { BaseAddress = new Uri(DefaultBaseUrl) };
    private readonly StashSession _session;

    public StashApi(StashSession session)
    {
        _session = session;
    }

    public Task<Whoami> WhoamiAsync(string? apiKeyOverride = null, CancellationToken ct = default) =>
        GetAsync<Whoami>("/api/whoami", apiKeyOverride, ct);

    public Task<ItemsPage> GetItemsAsync(int page = 1, int perPage = 48, string? search = null, string? type = null, CancellationToken ct = default)
    {
        var query = $"?page={page}&perPage={perPage}";
        if (!string.IsNullOrWhiteSpace(search)) query += $"&search={Uri.EscapeDataString(search)}";
        if (!string.IsNullOrWhiteSpace(type)) query += $"&type={Uri.EscapeDataString(type)}";
        return GetAsync<ItemsPage>($"/api/items{query}", null, ct);
    }

    private async Task<T> GetAsync<T>(string path, string? apiKeyOverride, CancellationToken ct)
    {
        var key = apiKeyOverride ?? _session.ApiKey
            ?? throw new StashApiException("Not signed in.");

        using var req = new HttpRequestMessage(HttpMethod.Get, path);
        req.Headers.Add("X-Api-Key", key);

        using var response = await _http.SendAsync(req, ct);
        if (response.StatusCode == HttpStatusCode.Unauthorized)
            throw new StashApiException("That API key was rejected. Check it in the Windows app under Settings > Cloud Sync > API Keys.");
        if (!response.IsSuccessStatusCode)
            throw new StashApiException($"The stash service returned an error (HTTP {(int)response.StatusCode}).");

        return await response.Content.ReadFromJsonAsync<T>(JsonOptions, ct)
            ?? throw new StashApiException("The stash service returned an empty response.");
    }
}
